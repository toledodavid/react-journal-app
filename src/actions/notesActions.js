import Swal from 'sweetalert2';
import { types } from '../types/types';
import { db } from '../firebase/firebase-config';
import { loadNotes } from '../helpers/loadNotes';
import { fileUpload } from '../helpers/fileUpload';


export const startNewEntryNoteAction = () => {
  return async (dispatch, getState) => {
    const {uid} = getState().auth;
    
    const newNote = {
      title: '',
      body: '',
      date: new Date().getTime()
    }

    const documentReference = await db.collection(`${uid}/journal/notes`).add(newNote);
    //console.log(documentReference);

    dispatch(activeNoteAction(documentReference.id, newNote));
    dispatch(addNewNoteAction(documentReference.id, newNote));
  }
}

export const activeNoteAction = (id, note) => ({
  type: types.noteActive,
  payload: {
    id,
    ...note
  }
});

export const addNewNoteAction = (idNote, note) => ({
  type: types.notesAddNewEntry,
  payload: {
    id: idNote,
    ...note
  }
});

export const startLoadingNotesAction = (uid) => {
  return async (dispatch) => {
    const notes = await loadNotes(uid);
    dispatch(setNotesAction(notes));
  }
}

export const setNotesAction = (notes) => ({
  type: types.notesLoad,
  payload: notes
});

export const startSaveNoteAction = (note) => {
  return async (dispatch, getState) => {
    const {uid} = getState().auth;

    if (!note.url) {
      delete note.url;
    }

    const noteToFirestore = {...note};
    delete noteToFirestore.id;

    await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore);

    dispatch(refreshNoteAction(note.id, note));
    
    Swal.fire('Saved', note.title, 'success');
  }
}

export const refreshNoteAction = (idNote, note) => ({
  type: types.noteUpdated,
  payload: {
    id: idNote,
    note
  }
});

export const startLoadingImageAction = (file) => {
  return async (dispatch, getState) => {
    const {active:note} = getState().notes;

    Swal.fire({
      title: 'Uploading',
      html: 'Please wait...',
      allowOutsideClick: false,
      didOpen: () => {
        Swal.showLoading();
      }
    });

    const fileUrl = await fileUpload(file);

    note.url = fileUrl;
    dispatch(startSaveNoteAction(note));

    Swal.close();
  }
}

export const startDeletingNoteAction = (idNote) => {
  return async (dispatch, getState) => {
    const {uid} = getState().auth;

    await db.doc(`${uid}/journal/notes/${idNote}`).delete();

    dispatch(deleteNoteAction(idNote));
  }
}

export const deleteNoteAction = (idNote) => ({
  type: types.noteDelete,
  payload: idNote
});

export const notesLogoutAction = () => ({
  type: types.notesLogoutCleaning
});