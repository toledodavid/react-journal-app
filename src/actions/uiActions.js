import { types } from '../types/types';


export const setErrorAction = (msgError) => ({
  type: types.uiSetError,
  payload: msgError
});

export const removeErrorAction = () => ({
  type: types.uiRemoveError
});

export const startLoadingAction = () => ({
  type: types.uiStartLoading
});

export const finishLoadingAction = () => ({
  type: types.uiFinishLoading
});