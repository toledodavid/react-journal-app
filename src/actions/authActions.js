import { firebase, googleAuthProvider } from '../firebase/firebase-config';
import { types } from '../types/types';
import { finishLoadingAction, startLoadingAction } from './uiActions';


export const startLoginEmailPassword = (email, password) => {
  return (dispatch) => {

    dispatch(startLoadingAction());

    firebase.auth().signInWithEmailAndPassword(email, password).then(({user}) => {
      dispatch(loginAction(user.uid, user.displayName));
      dispatch(finishLoadingAction());
    }).catch(err => {
      console.log(err);
      dispatch(finishLoadingAction());
    });
  }
}

export const startRegisterWithEmailPasswordNameAction = (email, password, name) => {
  return (dispatch) => {

    dispatch(startLoadingAction());

    firebase.auth().createUserWithEmailAndPassword(email, password).then( async({user}) => {
      await user.updateProfile({displayName: name});
      dispatch(loginAction(user.uid, user.displayName));
      dispatch(finishLoadingAction());
    }).catch(err => {
      console.log(err);
      dispatch(finishLoadingAction());
    });
  }
}

export const startGoogleLoginAction = () => {
  return (dispatch) => {
    firebase.auth().signInWithPopup(googleAuthProvider).then(({user}) => {
      dispatch(loginAction(user.uid, user.displayName));
    });
  }
}

export const loginAction = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    displayName
  }
});


export const startLogoutAction = () => {
  return async(dispatch) => {
    await firebase.auth().signOut();
    dispatch(logoutAction());
  }
}

export const logoutAction = () => ({
  type: types.logout
});