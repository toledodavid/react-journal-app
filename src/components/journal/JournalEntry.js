import React from 'react';
import moment from 'moment';
import { useDispatch } from 'react-redux';
import { activeNoteAction } from '../../actions/notesActions';




const JournalEntry = ({id, title, body, date, url}) => {

  const dispatch = useDispatch();

  const noteDate = moment(date);

  const handleEntryClick = () => {
    dispatch(activeNoteAction(id, {title, body, url, date}));
  }


  return(
    <div className="journal__entry pointer" onClick={handleEntryClick}>
      
      {
        url &&
          (
            <div className="journal__entry-picture"
                style={{backgroundSize: 'cover', backgroundImage: `url(${url})`}}>
            </div>
          )
      }

      <div className="journal__entry-body">
        <p className="jounral__entry-title">{title}</p>
        <p className="jounral__entry-content">{body}</p>
      </div>

      <div className="journal__entry-date-box">
        <span>{noteDate.format('dddd')}</span>
        <h4>{noteDate.format('Do')}</h4>
      </div>
    </div>
  );
}



export default JournalEntry;