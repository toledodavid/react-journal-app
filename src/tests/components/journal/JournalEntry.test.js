import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { activeNoteAction } from '../../../actions/notesActions';

import JournalEntry from '../../../components/journal/JournalEntry';


const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {};

let store = mockStore(initState);
store.dispatch = jest.fn();

const note = {
  id: 10,
  date: 0,
  title: 'Hello',
  body: 'world',
  url: 'https://somepage.com/foro.jpg'
};


describe('Tests in <JournalEntry /> component', () => {

  const wrapper = mount(
    <Provider store={store}>
      <JournalEntry {...note} />
    </Provider>
  );

  test('It should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('It should active note', () => {
    wrapper.find('.journal__entry').prop('onClick')();

    expect(store.dispatch).toHaveBeenCalled();
    expect(store.dispatch).toHaveBeenCalledWith(activeNoteAction(note.id, {...note}));
  });

});