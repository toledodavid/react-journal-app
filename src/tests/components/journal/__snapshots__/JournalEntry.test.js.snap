// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <JournalEntry /> component It should render correctly 1`] = `
<div
  className="journal__entry pointer animate__animated animate__fadeInUp"
  onClick={[Function]}
>
  <div
    className="journal__entry-picture"
    style={
      Object {
        "backgroundImage": "url(https://somepage.com/foro.jpg)",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="journal__entry-body"
  >
    <p
      className="jounral__entry-title"
    >
      Hello
    </p>
    <p
      className="jounral__entry-content"
    >
      world
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
      Wednesday
    </span>
    <h4>
      31st
    </h4>
  </div>
</div>
`;
